### PR TITLE
Version 2.0

### PR DESCRIPTION
**Pull request by tomneedham**
Rewrite of the editors js file, as started at the Nürnberg hack week. The main differences are that the code is rewritten inside the OCA.Files_Texteditor namespace. Functionality is very similar, with the exception of autosave. More complete changelog below:

 - Works in different file views
 - Now a popup, not some hack on the files UI (fullscreen for narrow/mobile UI)
 - Autosave
 - Loading feedback
 - Responsive (optimised for mobile usage)
 - Search removed in favour of native search functions (mobile and desktop build in browser features)

Bugs:
 - [x] Cursor moves past end of the visible window

Screenshot of the current state:
screen shot 2015-08-28 at 15 43 41
screen shot 2015-08-28 at 15 43 28

### CODE REVIEW

**Comment by tomneedham** [comment]
@jancborchardt

File: js/editor.js [hunk 1]
```diff
@@ -1,425 +1,453 @@
...
+				// Oh dear
+				OC.dialogs.alert(message, t('files_texteditor', 'An error occurred!'));
+			});
+	},
+
+	/**
+	 * Load the editor control bar
+	 */
+	loadControlBar: function(file, context) {
+		var html = '<small class="filename">'+file.name+'</small><button id="editor_save">' + t('files_texteditor', 'Save') + '</button><small class="lastsaved">'+t('files_texteditor', 'Last saved: never')+'</small>';
```
Comment (LukasReschke) [inline js/editor.js]: :bomb: :bomb: :bomb: :bomb: :bomb: 

`file.name` is not escaped

:bomb: :bomb: :bomb: :bomb: :bomb:

**Comment by DeepDiver1975** [comment]
@tomneedham needs rebase - THX

furthermore I'd like to ask you to do this in baby steps - a WIP-PR with more then 60k SLOC is never going to be reviewed properly

**Comment by tomneedham** [comment]
Ah the 60k will be the ace version update. May well be easier to leave that out to make reviewing on github easier.

The reason I didn't do this in baby steps (until now) is that I completely rewrote the javascript using objects so the code is pretty hard to compare to previous versions. But yes, I should get this to at least the current state of features and get that in before adding more fluff!

**Comment by DeepDiver1975** [comment]
> Ah the 60k will be the ace version update.

This is basically this commit - right? https://github.com/owncloud/files_texteditor/commit/a4827175ae19635316e6694b76631047bb1925ff

Maybe update ace in it's own PR?

**Comment by tomneedham** [comment]
> Maybe update ace in it's own PR?

That's the plan

**Comment by jancborchardt** [comment]
Great stuff here man! Looks nice with the overlay, and just showing the filename instead of the clunky full path. Good calls there! Will testdrive it too.

**Comment by jancborchardt** [comment]
@tomneedham pushed some design fixes. One important thing which doesn’t work right now is that it should of course automatically save when you close the document! :)

**Comment by schiesbn** [comment]
> pushed some design fixes. One important thing which doesn’t work right now is that it should of course automatically save when you close the document!

How can I close the document without saving then? I didn't tried it yet but on the screen shot I see a save button and a "X". My assumption would be that the X closes the window without saving.

**Comment by tomneedham** [comment]
I imagined it would maintain the current functionaly:
 - Press 'X'
 - Editor closes, without saving
 - Offers notification with a chance to resume editing

**Comment by jancborchardt** [comment]
Every last-generation editor asks you if you want to save, before you quit. The current-gen editors (like Sublime) just keep the saved changes inside the editor.

Closing the editor to get rid of all changes is a really bad outdated system. Autosave and the ability to undo (we have versions, folks) surpassed that in many apps. We should do so too.

**Comment by tomneedham** [comment]
@jancborchardt Question regarding the search box. Currently it makes sense to reuse the main nav bar search when in mobile mode as this takes up less space as is more clearly visible. However, it's less obvious when the editor is open in desktop browsers. How do we handle this? There is very little space on mobile for a search box within the editor control bar. Do we ignore searching on mobile devices? And before you say it, using the main nav search on mobile and a custom search on desktop is going to be extremely messy!

**Comment by tomneedham** [comment]
@jancborchardt Beginning to think we just stick with native browser find functionality? Much better implemented in mobile and those who need it know where to find it on desktop.

**Comment by tomneedham** [comment]
@jancborchardt Can you take a look at the issue with the text area extending below the visible region (type lots of lines until you reach the bottom) and also on Firefox the unsaved star is in the wrong place (strangely works on Chrome)

**Comment by jancborchardt** [comment]
Some first review:
* [ ] increase font-size to 13px
* [ ] put the »unsaved« asterisk right next to the filename, not right next to the unsaved msg
* [ ] only use grey for the row highlight instead of yellow, `#f8f8f8` please.
* [ ] what’s that glitch in the top right? Firefox

capture du 2015-08-30 12-40-45

**Comment by tomneedham** [comment]
@jancborchardt Retest please.

**Comment by tomneedham** [comment]
@MorrisJobke You too :)

**Comment by tomneedham** [comment]
and @icewind1991

**Comment by tomneedham** [comment]
@SergioBertolinSG

**Comment by icewind1991** [comment]
:+1:

**Comment by tomneedham** [comment]
@DeepDiver1975 also.

**Comment by MorrisJobke** [comment]
Really awesome :+1: I fixed the icon stuff for IE8 and it even works there

**Comment by SergioBertolinSG** [comment]
In safari the highlighted text when searching for a string is moved one character to the right.
Chrome and firefox are working well.
IE 8 doesn't work for me. It breaks the browser.

**Comment by jancborchardt** [comment]
Let’s merge this before it gets too big. Any small issues we should do in follow-up pull requests. :+1: great work @tomneedham!

**Pull request merged**